lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+    unsigned MaskDwords =
+        llvm::RoundUpToAlignment(NumPCOrPrimSigScalars, 32) / 32;
+    if (UsesViewID) {
+      ViewIDPCOutputMask = {&(Data[Offset]), MaskDwords};
+      Offset += MaskDwords;
+    }
+    unsigned TabSize = MaskDwords * NumInputSigScalars;
+    if (TabSize) {
+      unsigned AlignedTabSize =
+          MaskDwords * llvm::RoundUpToAlignment(NumInputSigScalars, 4);
+      InputToPCOutputTable = std::vector<uint32_t>(AlignedTabSize, 0);
+      memcpy(InputToPCOutputTable.data(), &(Data[Offset]), TabSize * 4);
+      Offset += TabSize;
+    }
+  } else if (SK == PSVShaderKind::Domain) {
+    // #PatchConstant.
+    NumPCOrPrimSigScalars = Data[Offset++];
+    unsigned OutputScalars = NumOutputSigScalars[0];
+    unsigned MaskDwords = llvm::RoundUpToAlignment(OutputScalars, 32) / 32;
+    unsigned TabSize = MaskDwords * NumPCOrPrimSigScalars;
+    if (TabSize) {
+      unsigned AlignedTabSize =
+          MaskDwords * llvm::RoundUpToAlignment(NumPCOrPrimSigScalars, 4);
+      PCInputToOutputTable = std::vector<uint32_t>(AlignedTabSize, 0);
+      memcpy(PCInputToOutputTable.data(), &(Data[Offset]), TabSize * 4);
+      Offset += TabSize;
+    }
+  }
+  IsValid = Offset == Data.size();
+}
+
+class PSVContentVerifier {
+  DxilModule &DM;
+  DxilPipelineStateValidation &PSV;
+  ValidationContext &ValCtx;
+  bool PSVContentValid = true;
+
+public:
+  PSVContentVerifier(DxilPipelineStateValidation &PSV, DxilModule &DM,
+                     ValidationContext &ValCtx)
+      : DM(DM), PSV(PSV), ValCtx(ValCtx) {}
+  void Verify();
+
+private:
+  void VerifySignatures(unsigned ValMajor, unsigned ValMinor);
+  void VerifySignature(const DxilSignature &, PSVSignatureElement0 *Base,
+                       unsigned Count, std::string Name,
+                       bool i1ToUnknownCompat);
+  void VerifySignatureElement(const DxilSignatureElement &,
+                              PSVSignatureElement0 *, const PSVStringTable &,
+                              const PSVSemanticIndexTable &, std::string, bool);
+  void VerifyResources(unsigned PSVVersion);
+  template <typename T>
+  void VerifyResourceTable(T &ResTab, unsigned &ResourceIndex,
+                           unsigned PSVVersion);
+  void VerifyViewIDDependence(PSVRuntimeInfo1 *PSV1);
+  void VerifyEntryProperties(const ShaderModel *SM, PSVRuntimeInfo0 *PSV0,
+                             PSVRuntimeInfo1 *PSV1, PSVRuntimeInfo2 *PSV2);
+  void EmitMismatchError(StringRef Name, StringRef PartContent,
+                         StringRef ModuleContent) {
+    ValCtx.EmitFormatError(ValidationRule::ContainerContentMatches,
+                           {Name, "PSV0", PartContent, ModuleContent});
+    PSVContentValid = false;
+  }
+  void EmitInvalidError(StringRef Name) {
+    ValCtx.EmitFormatError(ValidationRule::ContainerContentInvalid, {Name});
+    PSVContentValid = false;
+  }
+  template <typename Ty> static std::string GetDump(const Ty &T) {
```

Comment:
Just a note, though this is a full template, it's only ever used to output BindInfo0 and its subclass BindInfo1